README.md
# Rari Yield Pool: Smart Contracts

Welcome to `rari-yield-pool-contracts`, the central repository for the Solidity source code behind the Rari Yield Pool's Ethereum-based smart contracts (with automated tests and documentation).

## How it works

The Rari Yield Pool is a decentralized and fully-audited stablecoin lending aggregator optimized for yield based on the Ethereum blockchain. A high-level overview of how the Rari Yield Pool works is available in [`CONCEPT.md`](CONCEPT.md). This information is also [available online](https://rari.capital/current.html). Find out more about Rari Capital at [rari.capital](https://rari.capital).

## Contract usage

Documentation on common usage of the contracts is available in [`USAGE.md`](USAGE.md). Detailed API documentation for our smart contracts' public methods is available in [`API.md`](API.md). Smart contract ABIs are available in the `abi` properties of the JSON files in the `build` folder. For easy implementation, see the [Rari JavaScript SDK](https://github.com/Rari-Capital/rari-sdk).

## Installation (for development and deployment)

We, as well as others, had success using Truffle on Node.js `v12.18.2` with the latest version of NPM.

To install the latest version of Truffle: `npm install -g truffle`

*Though the latest version of Truffle should work, to compile, deploy, and test our contracts, we used Truffle `v5.1.45` (which should use `solc` version `0.5.17+commit.d19bba13.Emscripten.clang` and Web3.js `v1.2.1`).*

To install all our dependencies: `npm install`

## Compiling the contracts

`npm run compile`

## Testing the contracts

In `.env`, set `DEVELOPMENT_ADDRESS=0x45D54B22582c79c8Fb8f4c4F2663ef54944f397a` to test deployment and also set `DEVELOPMENT_ADDRESS_SECONDARY=0x1Eeb75CFad36EDb6C996f7809f30952B0CA0B5B9` to run automated tests.

If you are upgrading from `v1.2.0` or `v1.2.1`, set `UPGRADE_FROM_LAST_VERSION=1` to enable upgrading and configure the following:

    UPGRADE_FUND_OLD_CONTROLLER_ADDRESS=0x6afE6C37bF75f80D512b9D89C19EC0B346b09a8d
    UPGRADE_FUND_MANAGER_ADDRESS=0x59FA438cD0731EBF5F4cDCaf72D4960EFd13FCe6
    UPGRADE_FUND_OWNER_ADDRESS=0x10dB6Bce3F2AE1589ec91A872213DAE59697967a

Then, copy the OpenZeppelin artifacts for the official deployed `v1.2.0`/`v1.2.1` contracts from `.openzeppelin/mainnet.json` to `.openzeppelin/unknown-1337.json`. If you decide to disable upgrading by setting restoring `UPGRADE_FROM_LAST_VERSION=0`, make sure to delete `.openzeppelin/unknown-1337.json`.

To test the contracts, first fork the Ethereum mainnet. Begin by configuring `DEVELOPMENT_WEB3_PROVIDER_URL_TO_BE_FORKED` in `.env` (set to any mainnet Web3 HTTP provider JSON-RPC URL; we use a local `geth` instance, specifically a light client started with `geth --syncmode light --rpc --rpcapi eth,web3,debug,net`; Infura works too, but beware of latency and rate limiting). To start the fork, run `npm run ganache`. *If you would like to change the port, make sure to configure `scripts/ganache.js`, `scripts/test.sh`, and the `development` network in `truffle-config.js`.* Note that you will likely have to regularly restart your fork, especially when forking from a node without archive data or when using live 0x API responses to make currency exchanges.

To deploy the contracts to your private mainnet fork: `truffle migrate --network development --skip-dry-run --reset`

To run automated tests on the contracts on your private mainnet fork, run `npm test` (which runs `npm run ganache` in the background for you). If you are upgrading from `v1.2.0` or `v1.2.1`, you must also set the following variables in `.env`:

    UPGRADE_FUND_TOKEN_ADDRESS=0x3baa6B7Af0D72006d3ea770ca29100Eb848559ae
    UPGRADE_FUND_PRICE_CONSUMER_ADDRESS=0x00815e0e9d118769542ce24be95f8e21c60e5561
    UPGRADE_FUND_PROXY_ADDRESS=0x35DDEFa2a30474E64314aAA7370abE14c042C6e8

If you'd like to test gasless deposits via `RariFundProxy.deposit` via the Gas Station Network:

* Download `https://github.com/OpenZeppelin/openzeppelin-gsn-provider/blob/master/bin/gsn-relay` to `bin/gsn-relay` and set permissions with `chmod +x bin/gsn-relay`.
* Making sure `npx` is installed, run `npm dev-gsn`.
* Fund `RariFundProxy` using `npx @openzeppelin/gsn-helpers fund-recipient --recipient $RARI_FUND_PROXY_ADDRESS -n http://localhost:8546 -f $FROM_ADDRESS` or [this tool](https://www.opengsn.org/recipients) (or manually send ETH to `RelayHub(0xD216153c06E857cD7f72665E0aF1d7D82172F494).depositFor(address target)`).
* Run `rari-gsn-signer` with `pm2 start ecosystem.config.js` after configuring `ecosystem.config.js`.

## Live deployment

In `.env`, configure `LIVE_DEPLOYER_ADDRESS`, `LIVE_DEPLOYER_PRIVATE_KEY`, `LIVE_WEB3_PROVIDER_URL`, `LIVE_GAS_PRICE` (ideally, use the "fast" price listed by [ETH Gas Station](https://www.ethgasstation.info/)), `LIVE_FUND_OWNER`, `LIVE_FUND_REBALANCER`, `LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY`, `LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY`, and `LIVE_FUND_GSN_TRUSTED_SIGNER` to deploy to the mainnet.

If you are upgrading from `v1.2.0` or `v1.2.1`, set `UPGRADE_FROM_LAST_VERSION=1` to enable upgrading and configure the following:

    UPGRADE_FUND_OLD_CONTROLLER_ADDRESS=0x6afE6C37bF75f80D512b9D89C19EC0B346b09a8d
    UPGRADE_FUND_MANAGER_ADDRESS=0x59FA438cD0731EBF5F4cDCaf72D4960EFd13FCe6
    UPGRADE_FUND_OWNER_ADDRESS=0x10dB6Bce3F2AE1589ec91A872213DAE59697967a

You must also set `LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY` and `LIVE_UPGRADE_TIMESTAMP_COMP_CLAIMED` (set to current timestamp after claiming COMP awarded to the `RariFundController`; you should run migrations within 1 hour of this timestamp).

Then, migrate: `truffle migrate --network live`

If you'd like to provide gasless deposits via `RariFundProxy.deposit` via the Gas Station Network:

* Fund `RariFundProxy` using `npx @openzeppelin/gsn-helpers fund-recipient --recipient $RARI_FUND_PROXY_ADDRESS -n $ETHEREUM_NODE_URL -f $FROM_ADDRESS` or [this tool](https://www.opengsn.org/recipients) (or manually send ETH to `RelayHub(0xD216153c06E857cD7f72665E0aF1d7D82172F494).depositFor(address target)`).
* Run `rari-gsn-signer` with `pm2 start ecosystem.config.js --env production` after configuring `ecosystem.config.js`.

## License

See `LICENSE`.

## Credits

Rari Capital's smart contracts are developed by [David Lucid](https://github.com/davidlucid). Find out more about Rari Capital at [rari.capital](https://rari.capital).


CHANGELOG.md
# Changelog

## `v1.3.0` (contracts deployed 2020-02-14; all code pushed 2020-02-14)

* Upgraded mStable SAVE V1 to V2 (and implemented MTA rewards).
* Fixed bug in calculation of `outputFilledAmountUsd` in `RariFundController.marketSell0xOrdersFillOrKill`.

## `v1.2.1` (no contracts deployed; all code pushed 2020-12-07)

* Updated mStable swap tests to skip removed bAssets (i.e., DAI).

## `v1.2.0` (contracts deployed 2020-11-25; all code pushed 2020-11-29)

* Fixed bug in validation of mUSD redemption in `RariFundController.withdrawAndExchange`.

## `v1.1.0` (contracts deployed 2020-11-19; all code pushed 2020-11-29)

* Added 0.5% withdrawal fee (removing `RariFundManager.withdrawFees` to save gas).

## `v1.0.0` (contracts deployed 2020-10-20; all code pushed 2020-10-30)

* Copied Rari Stable Pool `v2.0.0` at commit [8a0405e](https://github.com/Rari-Capital/rari-stable-pool-contracts/commit/8a0405e8dacf908c79b3fe2999c153f81fbb5108).
* Implemented liquidity mining of RGT (Rari Governance Token) distributions.
* Rebranded to Rari Yield Pool and Rari Yield Pool Token (RYPT).
* Implemented lending via yVaults.
* Move dApp to `rari-dapp-legacy`. 
* Completely removed account balance limit functions.
* Minor improvements to contracts (e.g., replaced `disableFund` and `enableFund` with `setFundDisabled`).


test/5_fund_user.js
/**
 * COPYRIGHT © 2020 RARI CAPITAL, INC. ALL RIGHTS RESERVED.
 * Anyone is free to integrate the public (i.e., non-administrative) application programming interfaces (APIs) of the official Ethereum smart contract instances deployed by Rari Capital, Inc. in any application (commercial or noncommercial and under any license), provided that the application does not abuse the APIs or act against the interests of Rari Capital, Inc.
 * Anyone is free to study, review, and analyze the source code contained in this package.
 * Reuse (including deployment of smart contracts other than private testing on a private network), modification, redistribution, or sublicensing of any source code contained in this package is not permitted without the explicit permission of David Lucid of Rari Capital, Inc.
 * No one is permitted to use the software for any purpose other than those allowed by this license.
 * This license is liable to change at any time at the sole discretion of David Lucid of Rari Capital, Inc.
 */

const erc20Abi = require('./abi/ERC20.json');

const currencies = require('./fixtures/currencies.json');
const pools = require('./fixtures/pools.json');

const RariFundController = artifacts.require("RariFundController");
const RariFundManager = artifacts.require("RariFundManager");
const RariFundToken = artifacts.require("RariFundToken");
const RariFundPriceConsumer = artifacts.require("RariFundPriceConsumer");

if (parseInt(process.env.UPGRADE_FROM_LAST_VERSION) > 0) {
  RariFundManager.address = process.env.UPGRADE_FUND_MANAGER_ADDRESS;
  RariFundToken.address = process.env.UPGRADE_FUND_TOKEN_ADDRESS;
  RariFundPriceConsumer.address = process.env.UPGRADE_FUND_PRICE_CONSUMER_ADDRESS;
}

// These tests expect the owner and the fund rebalancer of RariFundManager to be set to process.env.DEVELOPMENT_ADDRESS
contract("RariFundManager, RariFundController", accounts => {
  it("should deposit to the fund, approve and deposit to pools, accrue interest, and withdraw from the fund", async () => {
    let fundControllerInstance = await RariFundController.deployed();
    let fundManagerInstance = await (parseInt(process.env.UPGRADE_FROM_LAST_VERSION) > 0 ? RariFundManager.at(process.env.UPGRADE_FUND_MANAGER_ADDRESS) : RariFundManager.deployed());
    let fundTokenInstance = await (parseInt(process.env.UPGRADE_FROM_LAST_VERSION) > 0 ? RariFundToken.at(process.env.UPGRADE_FUND_TOKEN_ADDRESS) : RariFundToken.deployed());
    let fundPriceConsumerInstance = await (parseInt(process.env.UPGRADE_FROM_LAST_VERSION) > 0 ? RariFundPriceConsumer.at(process.env.UPGRADE_FUND_PRICE_CONSUMER_ADDRESS) : RariFundPriceConsumer.deployed());

    // Get currency prices in USD used by contracts
    var currencyPricesInUsd = await fundPriceConsumerInstance.getCurrencyPricesInUsd.call();

    // For each currency of each pool
    for (const poolName of Object.keys(pools)) for (const currencyCode of Object.keys(pools[poolName].currencies)) {
      var amountBN = web3.utils.toBN(10 ** (currencies[currencyCode].decimals - 1));
      var amountUsdBN = amountBN.mul(currencyPricesInUsd[Object.keys(currencies).indexOf(currencyCode)]).div(web3.utils.toBN(10 ** currencies[currencyCode].decimals));
      
      // Check balances
      let initialAccountBalance = await fundManagerInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      let initialFundBalance = await fundManagerInstance.getFundBalance.call();
      let initialRftBalance = await fundTokenInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      
      // Approve tokens to RariFundManager
      var erc20Contract = new web3.eth.Contract(erc20Abi, currencies[currencyCode].tokenAddress);
      await erc20Contract.methods.approve(RariFundManager.address, amountBN.toString()).send({ from: process.env.DEVELOPMENT_ADDRESS });

      // RariFundManager.deposit
      await fundManagerInstance.deposit(currencyCode, amountBN, { from: process.env.DEVELOPMENT_ADDRESS });

      // Check balances and interest
      let postDepositAccountBalance = await fundManagerInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      assert(postDepositAccountBalance.gte(initialAccountBalance.add(amountUsdBN).mul(web3.utils.toBN(999999)).div(web3.utils.toBN(1000000))));
      let postDepositFundBalance = await fundManagerInstance.getFundBalance.call();
      assert(postDepositFundBalance.gte(initialFundBalance.add(amountUsdBN).mul(web3.utils.toBN(999999)).div(web3.utils.toBN(1000000))));
      let postDepositRftBalance = await fundTokenInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      assert(postDepositRftBalance.gt(initialRftBalance));
      let postDepositInterestAccrued = await fundManagerInstance.getInterestAccrued.call();
      let postDepositPoolBalance = await fundControllerInstance.getPoolBalance.call(["dYdX", "Compound", "Aave", "mStable", "yVault"].indexOf(poolName), currencyCode);

      // Deposit to pool (using Compound as an example)
      // TODO: Ideally, deposit to pool via rari-fund-rebalancer
      await fundControllerInstance.approveToPool(["dYdX", "Compound", "Aave", "mStable", "yVault"].indexOf(poolName), currencyCode, poolName === "mStable" ? web3.utils.toBN(2).pow(web3.utils.toBN(256)).subn(1) : amountBN, { from: process.env.DEVELOPMENT_ADDRESS });
      await fundControllerInstance.depositToPool(["dYdX", "Compound", "Aave", "mStable", "yVault"].indexOf(poolName), currencyCode, amountBN, { from: process.env.DEVELOPMENT_ADDRESS });

      // Force accrue interest
      await new Promise(resolve => setTimeout(resolve, 1000));
      await web3.eth.sendTransaction({ from: process.env.DEVELOPMENT_ADDRESS, to: process.env.DEVELOPMENT_ADDRESS, value: 0 });

      // Check balances and interest after waiting for interest
      var requireInterestAccrual = ["DAI", "TUSD"].indexOf(currencyCode) >= 0;
      var acceptMarginOfError = (poolName == "dYdX" && currencyCode != "DAI") || poolName == "mStable" || poolName == "yVault";
      if (acceptMarginOfError) var usdMarginOfErrorBN = web3.utils.toBN(1e13);
      let preWithdrawalAccountBalance = await fundManagerInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      assert(preWithdrawalAccountBalance[requireInterestAccrual ? "gt" : "gte"](acceptMarginOfError ? postDepositAccountBalance.sub(usdMarginOfErrorBN) : postDepositAccountBalance));
      let preWithdrawalFundBalance = await fundManagerInstance.getFundBalance.call();
      assert(preWithdrawalFundBalance[requireInterestAccrual ? "gt" : "gte"](acceptMarginOfError ? postDepositFundBalance.sub(usdMarginOfErrorBN) : postDepositFundBalance));
      let preWithdrawalRftBalance = await fundTokenInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      assert(preWithdrawalRftBalance.eq(postDepositRftBalance));
      let preWithdrawalInterestAccrued = await fundManagerInstance.getInterestAccrued.call();
      assert(preWithdrawalInterestAccrued[requireInterestAccrual ? "gt" : "gte"](acceptMarginOfError ? postDepositInterestAccrued.sub(usdMarginOfErrorBN) : postDepositInterestAccrued));
      let preWithdrawalPoolBalance = await fundControllerInstance.getPoolBalance.call(["dYdX", "Compound", "Aave", "mStable", "yVault"].indexOf(poolName), currencyCode);
      assert(preWithdrawalPoolBalance[requireInterestAccrual ? "gt" : "gte"](acceptMarginOfError ? postDepositPoolBalance.add(amountBN).subn(10) : postDepositPoolBalance.add(amountBN)));

      // RariFundManager.withdraw
      var withdrawalAmountBN = web3.utils.BN.min(amountBN, preWithdrawalPoolBalance);
      await fundManagerInstance.withdraw(currencyCode, withdrawalAmountBN, { from: process.env.DEVELOPMENT_ADDRESS, nonce: await web3.eth.getTransactionCount(process.env.DEVELOPMENT_ADDRESS) });

      // TODO: Check balances and assert with post-interest balances
      let finalAccountBalance = await fundManagerInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      assert(finalAccountBalance.lt(preWithdrawalAccountBalance));
      let finalFundBalance = await fundManagerInstance.getFundBalance.call();
      assert(finalFundBalance.lt(preWithdrawalFundBalance));
      let finalRftBalance = await fundTokenInstance.balanceOf.call(process.env.DEVELOPMENT_ADDRESS);
      assert(finalRftBalance.lt(preWithdrawalRftBalance));
    }
  });
});


migrations/2_deploy_contracts.js
/**
 * COPYRIGHT © 2020 RARI CAPITAL, INC. ALL RIGHTS RESERVED.
 * Anyone is free to integrate the public (i.e., non-administrative) application programming interfaces (APIs) of the official Ethereum smart contract instances deployed by Rari Capital, Inc. in any application (commercial or noncommercial and under any license), provided that the application does not abuse the APIs or act against the interests of Rari Capital, Inc.
 * Anyone is free to study, review, and analyze the source code contained in this package.
 * Reuse (including deployment of smart contracts other than private testing on a private network), modification, redistribution, or sublicensing of any source code contained in this package is not permitted without the explicit permission of David Lucid of Rari Capital, Inc.
 * No one is permitted to use the software for any purpose other than those allowed by this license.
 * This license is liable to change at any time at the sole discretion of David Lucid of Rari Capital, Inc.
 */

const { deployProxy, upgradeProxy, admin } = require('@openzeppelin/truffle-upgrades');
require('dotenv').config();

var DydxPoolController = artifacts.require("./lib/pools/DydxPoolController.sol");
var CompoundPoolController = artifacts.require("./lib/pools/CompoundPoolController.sol");
var AavePoolController = artifacts.require("./lib/pools/AavePoolController.sol");
var MStablePoolController = artifacts.require("./lib/pools/MStablePoolController.sol");
var YVaultPoolController = artifacts.require("./lib/pools/YVaultPoolController.sol");
var ZeroExExchangeController = artifacts.require("./lib/exchanges/ZeroExExchangeController.sol");
var MStableExchangeController = artifacts.require("./lib/exchanges/MStableExchangeController.sol");
var RariFundController = artifacts.require("./RariFundController.sol");
var RariFundManager = artifacts.require("./RariFundManager.sol");
var RariFundToken = artifacts.require("./RariFundToken.sol");
var RariFundPriceConsumer = artifacts.require("./RariFundPriceConsumer.sol");
var RariFundProxy = artifacts.require("./RariFundProxy.sol");

module.exports = async function(deployer, network, accounts) {
  if (["live", "live-fork"].indexOf(network) >= 0) {
    if (!process.env.LIVE_GAS_PRICE) return console.error("LIVE_GAS_PRICE is missing for live deployment");
    if (!process.env.LIVE_FUND_OWNER) return console.error("LIVE_FUND_OWNER is missing for live deployment");
    if (!process.env.LIVE_FUND_REBALANCER) return console.error("LIVE_FUND_REBALANCER is missing for live deployment");
    if (!process.env.LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY) return console.error("LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY is missing for live deployment");
    if (!process.env.LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY) return console.error("LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY is missing for live deployment");
  }
  
  if (parseInt(process.env.UPGRADE_FROM_LAST_VERSION) > 0) {
    // Upgrade from v1.2.0 (only modifying RariFundController v1.0.0 and RariFundManager v1.1.0) to v1.3.0
    if (!process.env.UPGRADE_OLD_FUND_CONTROLLER) return console.error("UPGRADE_OLD_FUND_CONTROLLER is missing for upgrade");
    if (!process.env.UPGRADE_FUND_MANAGER_ADDRESS) return console.error("UPGRADE_FUND_MANAGER_ADDRESS is missing for upgrade");
    if (!process.env.UPGRADE_FUND_OWNER_ADDRESS) return console.error("UPGRADE_FUND_OWNER_ADDRESS is missing for upgrade");

    if (["live", "live-fork"].indexOf(network) >= 0) {
      if (!process.env.LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY) return console.error("LIVE_UPGRADE_FUND_OWNER_PRIVATE_KEY is missing for live upgrade");
      if (!process.env.LIVE_UPGRADE_TIMESTAMP_COMP_CLAIMED || process.env.LIVE_UPGRADE_TIMESTAMP_COMP_CLAIMED < ((new Date()).getTime() / 1000) - 3600 || process.env.LIVE_UPGRADE_TIMESTAMP_COMP_CLAIMED > (new Date()).getTime() / 1000) return console.error("LIVE_UPGRADE_TIMESTAMP_COMP_CLAIMED is missing, invalid, or out of date for live upgrade");
    } else {
      if (!process.env.UPGRADE_FUND_TOKEN_ADDRESS) return console.error("UPGRADE_FUND_TOKEN_ADDRESS is missing for development upgrade");
      if (!process.env.UPGRADE_FUND_PRICE_CONSUMER_ADDRESS) return console.error("UPGRADE_FUND_PRICE_CONSUMER_ADDRESS is missing for development upgrade");
      if (!process.env.UPGRADE_FUND_PROXY_ADDRESS) return console.error("UPGRADE_FUND_PROXY_ADDRESS is missing for development upgrade");
    }

    // Upgrade from v1.2.0 (RariFundController v1.0.0) to v1.3.0
    var oldRariFundController = await RariFundController.at(process.env.UPGRADE_OLD_FUND_CONTROLLER);

    // Deploy liquidity pool and currency exchange libraries
    await deployer.deploy(DydxPoolController);
    await deployer.deploy(CompoundPoolController);
    await deployer.deploy(AavePoolController);
    await deployer.deploy(MStablePoolController);
    await deployer.deploy(YVaultPoolController);
    await deployer.deploy(ZeroExExchangeController);
    await deployer.deploy(MStableExchangeController);

    // Link libraries to RariFundController
    await deployer.link(DydxPoolController, RariFundController);
    await deployer.link(CompoundPoolController, RariFundController);
    await deployer.link(AavePoolController, RariFundController);
    await deployer.link(MStablePoolController, RariFundController);
    await deployer.link(YVaultPoolController, RariFundController);
    await deployer.link(ZeroExExchangeController, RariFundController);
    await deployer.link(MStableExchangeController, RariFundController);

    // Deploy new RariFundController
    var rariFundController = await deployer.deploy(RariFundController);

    // Disable the fund on the old RariFundController
    await oldRariFundController.setFundDisabled(true, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });

    // Disable the fund on the RariFundManager and set fund controller to 0x0 (so getFundBalance reverts)
    RariFundManager.class_defaults.from = process.env.UPGRADE_FUND_OWNER_ADDRESS;
    var rariFundManager = await RariFundManager.at(process.env.UPGRADE_FUND_MANAGER_ADDRESS);
    await rariFundManager.setFundDisabled(true);
    await rariFundManager.setFundController("0x0000000000000000000000000000000000000000");

    // Upgrade RariFundController
    await oldRariFundController.methods["upgradeFundController(address)"](RariFundController.address, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });

    // Forward COMP governance tokens
    if (oldRariFundController.methods["upgradeFundController(address,address)"].call(RariFundController.address, "0xc00e94cb662c3520282e6f5717214004a7f26888", { from: process.env.UPGRADE_FUND_OWNER_ADDRESS })) await oldRariFundController.methods["upgradeFundController(address,address)"](RariFundController.address, "0xc00e94cb662c3520282e6f5717214004a7f26888", { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });

    // Check yVaults for funds
    for (const currencyCode of ["DAI", "USDC", "USDT", "TUSD"]) if (await rariFundController.checkPoolForFunds.call(4, currencyCode)) await rariFundController.checkPoolForFunds(4, currencyCode);

    // Connect new RariFundController and RariFundManager
    await rariFundController.setFundManager(process.env.UPGRADE_FUND_MANAGER_ADDRESS);
    await rariFundManager.setFundController(RariFundController.address);

    // Re-enable the fund on the RariFundManager
    await rariFundManager.setFundDisabled(false);

    // Set Aave referral code
    await rariFundController.setAaveReferralCode(86);

    // Set daily loss rate limit for currency exchanges
    await rariFundController.setDailyLossRateLimit(["live", "live-fork"].indexOf(network) >= 0 ? web3.utils.toBN(0.02e18) : web3.utils.toBN(0.9e18));

    // Set fund rebalancer on controller and manager
    await rariFundController.setFundRebalancer(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_REBALANCER : process.env.DEVELOPMENT_ADDRESS);

    if (["live", "live-fork"].indexOf(network) >= 0) {
      // Live network: transfer ownership of RariFundController to live owner
      await rariFundController.transferOwnership(process.env.LIVE_FUND_OWNER);
    } else {
      // Development network: transfer ownership of contracts to development address, set development address as rebalancer, and set all currencies to accepted
      await rariFundManager.transferOwnership(process.env.DEVELOPMENT_ADDRESS, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });
      var rariFundProxy = await RariFundProxy.at(process.env.UPGRADE_FUND_PROXY_ADDRESS); 
      await rariFundProxy.transferOwnership(process.env.DEVELOPMENT_ADDRESS, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });
      // TODO: await admin.transferProxyAdminOwnership(process.env.DEVELOPMENT_ADDRESS, { from: process.env.UPGRADE_FUND_OWNER_ADDRESS });
      RariFundManager.class_defaults.from = process.env.DEVELOPMENT_ADDRESS;
      await rariFundManager.setFundRebalancer(process.env.DEVELOPMENT_ADDRESS);
      await rariFundManager.setAcceptedCurrencies(["DAI", "USDC", "USDT", "TUSD", "BUSD", "sUSD", "mUSD"], [true, true, true, true, true, true, true]);
    }
  } else {
    // Normal deployment!
    // Deploy liquidity pool and currency exchange libraries
    await deployer.deploy(DydxPoolController);
    await deployer.deploy(CompoundPoolController);
    await deployer.deploy(AavePoolController);
    await deployer.deploy(MStablePoolController);
    await deployer.deploy(YVaultPoolController);
    await deployer.deploy(ZeroExExchangeController);
    await deployer.deploy(MStableExchangeController);

    // Link libraries to RariFundController
    await deployer.link(DydxPoolController, RariFundController);
    await deployer.link(CompoundPoolController, RariFundController);
    await deployer.link(AavePoolController, RariFundController);
    await deployer.link(MStablePoolController, RariFundController);
    await deployer.link(YVaultPoolController, RariFundController);
    await deployer.link(ZeroExExchangeController, RariFundController);
    await deployer.link(MStableExchangeController, RariFundController);

    // Deploy RariFundController and RariFundManager
    var rariFundController = await deployer.deploy(RariFundController);
    var rariFundManager = await deployProxy(RariFundManager, [], { deployer });

    // Connect RariFundController and RariFundManager
    await rariFundController.setFundManager(RariFundManager.address);
    await rariFundManager.setFundController(RariFundController.address);

    // Set Aave referral code
    await rariFundController.setAaveReferralCode(86);
    
    // Deploy RariFundToken
    var rariFundToken = await deployProxy(RariFundToken, [], { deployer });
    
    // Add RariFundManager as as RariFundToken minter
    await rariFundToken.addMinter(RariFundManager.address);

    // Connect RariFundToken to RariFundManager
    await rariFundManager.setFundToken(RariFundToken.address);

    // Deploy RariFundPriceConsumer
    var rariFundPriceConsumer = await deployProxy(RariFundPriceConsumer, [], { deployer });

    // Connect RariFundPriceConsumer to RariFundManager
    await rariFundManager.setFundPriceConsumer(RariFundPriceConsumer.address);

    // Set daily loss rate limit for currency exchanges
    await rariFundController.setDailyLossRateLimit(["live", "live-fork"].indexOf(network) >= 0 ? web3.utils.toBN(0.02e18) : web3.utils.toBN(0.9e18));

    // Set fund rebalancer on controller and manager
    await rariFundController.setFundRebalancer(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_REBALANCER : process.env.DEVELOPMENT_ADDRESS);
    await rariFundManager.setFundRebalancer(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_REBALANCER : process.env.DEVELOPMENT_ADDRESS);

    // Set interest fee master beneficiary
    await rariFundManager.setInterestFeeMasterBeneficiary(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_INTEREST_FEE_MASTER_BENEFICIARY : process.env.DEVELOPMENT_ADDRESS);

    // Set interest fee rate to 9.5%
    await rariFundManager.setInterestFeeRate(web3.utils.toBN(0.095e18));
  
    // Set withdrawal fee master beneficiary
    await rariFundManager.setWithdrawalFeeMasterBeneficiary(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_WITHDRAWAL_FEE_MASTER_BENEFICIARY : process.env.DEVELOPMENT_ADDRESS);
  
    // Set withdrawal fee rate to 0.5%
    await rariFundManager.setWithdrawalFeeRate(web3.utils.toBN(0.005e18));

    // Link libraries to RariFundProxy
    await deployer.link(ZeroExExchangeController, RariFundProxy);
    await deployer.link(MStableExchangeController, RariFundProxy);

    // Deploy RariFundProxy
    var rariFundProxy = await deployer.deploy(RariFundProxy);

    // Connect RariFundManager and RariFundProxy
    await rariFundManager.setFundProxy(RariFundProxy.address);
    await rariFundProxy.setFundManager(RariFundManager.address);

    // Set GSN trusted signer
    await rariFundProxy.setGsnTrustedSigner(["live", "live-fork"].indexOf(network) >= 0 ? process.env.LIVE_FUND_GSN_TRUSTED_SIGNER : process.env.DEVELOPMENT_ADDRESS);

    if (["live", "live-fork"].indexOf(network) >= 0) {
      // Live network: transfer ownership of deployed contracts from the deployer to the owner
      await rariFundController.transferOwnership(process.env.LIVE_FUND_OWNER);
      await rariFundManager.transferOwnership(process.env.LIVE_FUND_OWNER);
      await rariFundToken.addMinter(process.env.LIVE_FUND_OWNER);
      await rariFundToken.renounceMinter();
      await rariFundToken.addPauser(process.env.LIVE_FUND_OWNER);
      await rariFundToken.renouncePauser();
      await rariFundProxy.transferOwnership(process.env.LIVE_FUND_OWNER);
      await admin.transferProxyAdminOwnership(process.env.LIVE_FUND_OWNER);
    } else {
      // Development network: set all currencies to accepted
      await rariFundManager.setAcceptedCurrencies(["DAI", "USDC", "USDT", "TUSD", "BUSD", "sUSD", "mUSD"], [true, true, true, true, true, true, true]);
    }
  }
};
